Sage/carbon
# Focus trap fixes

### Proposed behaviour



Fixes 2 separate issues within the FocusTrap internal component (which is used by Dialog and related components to trap focus inside when tabbing):

1) Carbon should as far as possible not interfere with the behaviour of any 3rd-party libraries which the consumer is using that, by not calling `preventDefault` on the keyboard event (see first paragraph of "current behaviour" for details). This has been achieved by relying on focusable "tab guard" elements before and after the focus trap which "redirect" the focus as needed when focused, and otherwise relying on browser default behaviour (as modified by any other libraries in use) except when this would result in incorrect behaviour in Carbon (which happens in only one situation - when the consumer is using a custom `focusableSelectors` string - and even there the PR avoids calling `preventDefault` except when absolutely necessary). (fixes #5749 )

2) when multiple dialogs are open and a non-focusable element is clicked, resulting in focus being lost to the document body, tabbing should always result in focus going to the first focusable element of the top modal (fixes #5754 )

### Current behaviour

1) The current implementation relies on calling the `preventDefault` method on the keyboard event whenever the user tabs, in order to work out programatically the next element to focus. This breaks any third-party libraries - such as AG Grid - which implement their own behaviour on tab press and check that the default wasn't prevented before doing anything (thereby effectively making their implementation a new default, as calling `preventDefault` stops the library code from working as intended).

2) when multiple dialogs are open and a non-focusable element is clicked, resulting in focus being lost to the document body, tabbing can lead to a race condition between more than one focus trap both trying to focus their own first element. This leads to the possibility that the wrong element may be focused, as it is only the "top" modal that should get the focus on its first element.



### Checklist




#### QA


### Additional context



### Testing instructions


Ensure both linked issues have been fixed.

Also regression test all components that use FocusTrap - Dialog, DialogFullScree, Sidebar, MenuFullScreen and VerticalMenuFullScreen
